refactor(docs/plugins): update plugin documentation for 1.0.0 version changes (#581)
falcosecurity/falco-website: content/en/docs/plugins/_index.md

Plugins can be written in any language, as long as they export the required functions. Go, however, is the preferred language to write plugins, followed by C/C++.

> Plugin can implement one or more capabilities. In the scope of plugins, a *capability* is an extension of Falco's features in the form of a specific set of C function symbols exported by shared libraries. Currently, there are two plugin capabilities supported by the framework: *event sourcing* and *field extraction*.

Reviewer: ```suggestion
Plugins can implement one or more capabilities. In the scope of plugins, a *capability* is an extension of Falco's features in the form of a specific set of C function symbols exported by shared libraries. Currently, there are two plugin capabilities supported by the framework: *event sourcing* and *field extraction*.
```

### Plugins are Coresident with Falco

The libraries will do everything possible to validate the data coming from the plugins and protect Falco and the other consumers from corrupted data. However, for performance reasons, plugins are *"trusted"*: they run in the same thread and address space as Falco and they could crash the program. We assume that the user will be in control of plugin loading and will make sure only trusted plugins are loaded/packaged with Falco.

### Plugin SDKs

To make it easier to write plugins, there are [Go](https://github.com/falcosecurity/plugin-sdk-go) and [C++](https://github.com/falcosecurity/plugin-sdk-cpp) SDKs that handle the details of memory management and type conversion. These SDKs provide a streamlined way to implement plugins without having to deal with all the details of the lower-level functions that make up the Plugin API.

These SDKs are optional, but using them is highly recommended.

### Event Sourcing Capability

> Plugins with event sourcing capability provide a new event source and make it available to libscap and libsinsp. They have the ability to "open" and "close" a stream of events and return those events to the plugin framework. They also provide a plugin ID, which is globally unique and is used in capture files (see below). Event sources provided by plugins with this capability are tied to the events they generate, and can be used by plugins with field extraction capabilities and within Falco rules (see below).

Reviewer: ```suggestion
Plugins with event sourcing capability provide a new event source and make it available to libscap and libsinsp. They have the ability to "open" and "close" a stream of events and return those events to the plugin framework. They also provide a plugin ID, which is globally unique and is used in capture files (see below). Event sources provided by plugins with this capability are tied to the events they generate and can be used by plugins with field extraction capabilities and within Falco rules (see below).
```

### Field Extraction Capability

> Plugins with field extraction capability have the ability to extract information from events based on fields. For example, a field (e.g. `proc.name`) extracts a value (e.g. process name like `nginx`) from a syscall event. The plugin returns a set of supported fields, and there are functions to extract a value given an event and field. The plugin framework can then build filtering expressions/Falco rule conditions based on these fields combined with relational and/or logical operators. For example, given an expression `ct.name=root and ct.region=us-east-1`, the plugin framework handles parsing the expression, calling the plugin to extract values for fields `ct.name`/`ct.region` for a given event, and determining the result of the expression. In a Falco output string like `An EC2 Node was created (name=%ct.name region=%ct.region)`, the plugin framework handles parsing the output string, calling the plugin to extract values for fields, and building the resolved string, replacing the template field names (e.g. `%ct.region`) with values (e.g. `us-east-1`).

Reviewer: ```suggestion
Plugins with field extraction capability have the ability to extract information from events based on fields. For example, a field (e.g. `proc.name`) extracts a value (e.g. process name like `nginx`) from a syscall event. The plugin returns a set of supported fields, and there are functions to extract a value given an event and field. The plugin framework can then build filtering expressions (e.g. rule conditions) based on these fields combined with relational and/or logical operators. For example, given the expression `ct.name=root and ct.region=us-east-1`, the plugin framework handles parsing the expression, calling the plugin to extract values for fields `ct.name`/`ct.region` for a given event, and determining the result of the expression. In a Falco output string like `An EC2 Node was created (name=%ct.name region=%ct.region)`, the plugin framework handles parsing the output string, calling the plugin to extract values for fields, and building the resolved string, replacing the template field names (e.g. `%ct.region`) with values (e.g. `us-east-1`).
```

> Plugins with this capability focus only on field extraction from events generated by other plugins, or by the core libraries. They do **not** provide an event source, but can extract fields from other event sources. The supported field extraction can be generic or be tied to a specific event source. An example is json field extraction, where a plugin might be able to extract fields from generic json payloads.

Reviewer: ```suggestion
Plugins with this capability only focus on field extraction from events generated by other plugins or by the core libraries. They do **not** provide an event source but can extract fields from other event sources. The supported field extraction can be generic or be tied to a specific event source. An example is json field extraction, where a plugin might be able to extract fields from generic json payloads.
```

## Plugin Event IDs

Every plugin with event sourcing capability requires its own, unique plugin event ID to interoperate with Falco and the other plugins. This ID is used in the following ways:

- The ID is saved in in-memory event objects and is used to identify the associated plugin that injected the event.
- The ID is saved in capture files and is used to recreate in-memory event objects when reading capture files.

The ID must be unique to ensure that events written by a given plugin will be properly associated with that plugin (and its event sources, see below).

Plugin authors *must* register the plugin with the Falcosecurity organization by creating a PR to modify the [plugin registry](https://github.com/falcosecurity/plugins/blob/master/registry.yaml) with details on the new plugin. This ensures that a given ID is used by exactly one plugin.

## Plugin Event Sources and Interoperability

> Events returned by plugins with event sourcing capability have an "event source" that describes the event's information. This is distinct from the plugin name to allow for multiple plugin implementations to generate the same kind of events. For example, there might be plugins gke-k8saudit, eks-k8saudit, ibmcloud-k8saudit, etc. that all fetch [K8s Audit](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/) information. The plugins would have different names and IDs but would have the same event source `k8s_audit`.

Reviewer: ```suggestion
Events returned by plugins with event sourcing capability have an *event source* that describes the event's information. This is distinct from the plugin name to allow for multiple plugin implementations to generate the same kind of events. For example, there might be plugins gke-k8saudit, eks-k8saudit, ibmcloud-k8saudit, etc. that all fetch [K8s Audit](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/) information. The plugins would have different names and IDs but would have the same event source `k8s_audit`.
```

> A plugin with field extraction capability optionally provides a set of compatible event sources. When the framework receives an event with an event source in the plugin's set of event sources, fields in expressions/Falco outputs will be extracted from events using the plugin. The set of compatible event sources can also **be omitted**. In this case, *all* events will be presented to the plugin, regardless of their source. In this case, the plugin must detect the format of arbitrary payloads and be able to return NULL/*no value* when the payload is not supported. As such, given a specific event source such as `k8s_audit`, there is an implicit contract to honor regarding how data is formatted in each event of that source, such that compatible plugins with field extraction capability are able parse events of a certain source even if they are produced by different plugins.

Reviewer: ```suggestion
A plugin with field extraction capability optionally provides a set of compatible event sources. When the framework receives an event with an event source in the plugin's set of event sources, fields in expressions (i.e., fields included in the rule's output field) will be extracted from events using the plugin. The set of compatible event sources can also **be omitted**. In this case, *all* events will be presented to the plugin, regardless of their source. In this case, the plugin must detect the format of arbitrary payloads and be able to return NULL/*no value* when the payload is not supported. As such, given a specific event source such as `k8s_audit`, there is an implicit contract to honor regarding how data is formatted in each event of that source, such that compatible plugins with field extraction capability are able to parse events of a certain source even if they are produced by different plugins.
```

When managing plugins, we only need to ensure the following:

* That only one plugin is loaded at a time that exports a given event source. For example, the libraries can load either a gke-k8saudit plugin with event source `k8s_audit`, or eks-k8saudit with event source `k8s_audit`, but not both.
* That for a mix of plugins with event sourcing and field extraction capabilities having the same event source, that the fields are distinct. For example, a plugin event sourcing capability providing the source `k8s_audit` can export `ka.*` fields, and a plugin with field extractor capabilities with event source `k8s_audit` can export a `jevt.value[/...]` field, and the appropriate plugin will be used to extract fields from `k8s_audit` events as fields are parsed from condition expressions/output format strings.

## Plugin API

> Here is an overview of the functions that comprise the plugins API. This list is not extensive: the [plugin API reference](./plugin-api-reference) has full documentation of plugin APIs for all the capabilites supported by the framework.

Reviewer: ```suggestion
Here is an overview of the functions that comprise the plugin API. This list is not extensive: the [plugin API reference](./plugin-api-reference) has full documentation of plugin APIs for all the capabilities supported by the framework.
```

load_plugins: [cloudtrail, json]
```

A new `plugins` property in `falco.yaml` will define the set of plugins that can be loaded by Falco, and a new `load_plugins` property will control which plugins are actually loaded when Falco starts.

For more information, see [Configuration](../configuration).

The mechanics of loading a plugin are implemented in the libraries and leverage the dynamic library functionality of the operating system (dlopen/dlsym in unix, LoadLibrary/GetProcAddress in Windows). The plugin loading code also ensures that:

- the plugin is valid, i.e. that it exports the set of expected symbols
- the plugin has an api version number that is compatible with the plugin framework.
- that only one plugin with event sourcing capability is loaded at a time for a given event source
- if a mix of plugins for both event sourcing and field extraction are loaded for a given event source, that the exported fields have unique names that don't overlap across plugins

## Event Sources and Falco Rules

> Falco rules already have the notion of a "source", using the source property in YAML rules objects, and there is primarily one kind of event source: `syscall`. The source property in Falco rules maps a given rule to the event source on which the rule runs.

Reviewer: ```suggestion
Falco rules already have the notion of a *source*, using the source property in YAML rules objects, and there is primarily one kind of event source: `syscall`. The source property in Falco rules maps a given rule to the event source on which the rule runs.
```

For example, given a plugin providing events with source `aws_cloudtrail`, and a Falco rule with source property `aws_cloudtrail`, the rule will be evaluated for any events returned by the AWS Cloudtrail plugin.

> Similarly, an plugin with field extraction capability that includes `aws_cloudtrail` in its set of event sources will have the opportunity to extract information from Cloudtrail events. As a result, fields exported by the plugin can be put in a rule's condition, exception, or output properties when the rule has a source `aws_cloudtrail`.

Reviewer: ```suggestion
Similarly, a plugin with field extraction capability that includes `aws_cloudtrail` in its set of event sources will have the opportunity to extract information from Cloudtrail events. As a result, fields exported by the plugin can be put in a rule's condition, exception, or output properties when the rule has a source `aws_cloudtrail`.
```

## Plugins

> Plugins hosted and maintained by the Falcosecurity community are at the [plugins github repository](https://github.com/falcosecurity/plugins). There, you can also find the [plugin registry](https://github.com/falcosecurity/plugins/blob/master/registry.yaml) containing info about all the plugins officially recognized by the Falcosecurity organization, which include both the ones hosted by the community and the external ones.

Reviewer: ```suggestion
Plugins hosted and maintained by the Falcosecurity community are at the [plugins GitHub repository](https://github.com/falcosecurity/plugins). There, you can also find the [plugin registry](https://github.com/falcosecurity/plugins/blob/master/registry.yaml) containing info about all the plugins officially recognized by the Falcosecurity organization, which include both the ones hosted by the community and the external ones.
```

## Plugin SDKs

To facilitate the development of plugins written in Go, we've written a [SDK](https://github.com/falcosecurity/plugin-sdk-go) that provides support code for writing plugins. The SDK provides Go structs/enums corresponding to the C structs/enums used by the API, has utility packages that manage the details of memory management/type conversion, and presents abstract interfaces that provide a more streamlined interface to potential plugin authors. We go through the details and the architecture of the GO SDK in the [Go SDK walkthough section](./go-sdk-walkthrough).

There is also an experimental [C++](https://github.com/falcosecurity/plugins/tree/master/sdk/cpp) SDK that defines abstract C++ base classes for plugins. Plugin authors can derive from these base classes and implement the abstract methods to provide demographic information, events, and extract fields from events. This SDK is currently in line with the most recent changes in the plugin API, and will be subject to many updates in the near future.

All of the Falcosecurity-provided plugins use these SDKs.

## Falco

> Falco itself includes the k8saudit, cloudtrail and json plugins in its packages/container images. The plugins are defined in falco.yaml but by default no plugins are loaded when Falco starts.

Reviewer: ```suggestion
Falco itself includes the `k8saudit`, `cloudtrail` and `json` plugins in its packages and container images. The plugins are defined in falco.yaml but by default, no plugins are loaded when Falco starts.
```
